### PR TITLE
Results service

### PR DESCRIPTION
**Pull request by Chshravani**
This branch contains,
class service
subject service
schedule service
student service
results service

Collections involved in these APIs are class, results, schedule, student.

Testing- JUnit is used.
Testing is done for service and controller layer of class API, schedule API, subject API, results API.

class service, subject services are completely done.
schedule service, student service, results service are in progress.

### CODE REVIEW

File: src/main/java/com/target/targetreadyresultsservice/controller/ClassController.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    public ClassController(ClassService classService) {
+        this.classService = classService;
+    }
+
+    @GetMapping("/classes")
+    public ResponseEntity<List<ClassDto>> getClassDetails() {
+        try {
+            List<ClassDto> classes = classService.getAllClasses();
+            if (classes != null && classes.toString() !="") {
```
Comment (IamYashwant) [inline src/main/java/com/target/targetreadyresultsservice/controller/ClassController.java:29]: why this check on a list classes.toString() !="" ?
you can check for null and empty

File: src/main/java/com/target/targetreadyresultsservice/Exception/BlankValueException.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.target.targetreadyresultsservice.Exception;
+
+public class BlankValueException extends RuntimeException{
+    public BlankValueException(){
```
Comment (IamYashwant) [inline src/main/java/com/target/targetreadyresultsservice/Exception/BlankValueException.java:4]: you can put default messages to all exceptions.

**Comment by Chshravani** [comment]
1. This branch contains, 
    class service
    subject service
    schedule service
    student service
    results service

2. Collections involved in these APIs are class, results, schedule, student.

3. Testing- JUnit is used.
    Testing is done for service and controller layer of class API, schedule API, subject API, results API.

4. class service, subject services are  completely done.
    schedule service, student service, results service are in progress.

File: build.gradle [hunk 3]
```diff
@@ -20,9 +20,23 @@ dependencies {
...
+    testImplementation 'junit:junit:4.13.1'
+    testImplementation 'org.testng:testng:7.1.0'
+    testImplementation 'org.testng:testng:7.1.0'
+
     developmentOnly 'org.springframework.boot:spring-boot-devtools'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     testImplementation 'io.projectreactor:reactor-test'
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.1'
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.1'
+    implementation 'org.hibernate.validator:hibernate-validator:8.0.0.Final'
```
Comment (IamYashwant) [inline build.gradle:39]: group it accordingly. ex. all implementation together

File: src/main/java/com/target/targetreadyresultsservice/controller/ResultsController.java [hunk 4]
```diff
@@ -1,60 +1,111 @@
...
+        }
+    }
+
+
+    //get class results of particular test/exam
+    @GetMapping
+    public ResponseEntity<List<Results>> getClassTestResults(
+            @RequestParam("className") String className,
+            @RequestParam("academicYear") String acyear,
+             @RequestParam("scheudleName") String scName
```
Comment (IamYashwant) [inline src/main/java/com/target/targetreadyresultsservice/controller/ResultsController.java:53]: acyear 
follow the same convention like scName

File: src/main/java/com/target/targetreadyresultsservice/model/Subject.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+
+    public Subject(String subjectCode, @NotNull String subjectName, @NotNull Integer credits, @NotNull String classCode,@NotNull Integer maxExamMarks,@NotNull Integer maxTestMarks) {
+
+        this.subjectCode = subjectCode;
+        this.subjectName = subjectName;
+        this.credits = credits;
+        this.classCode = classCode;
+        this.maxExamMarks=maxExamMarks;
+        this.maxTestMarks=maxTestMarks;
```
Comment (IamYashwant) [inline src/main/java/com/target/targetreadyresultsservice/model/Subject.java:42]: care the formatting.

File: src/main/java/com/target/targetreadyresultsservice/service/ScheduleService.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        sc.setScheduleType(schedule.getScheduleType());
+        sc.setScheduleName(schedule.getScheduleName());
+        sc.setYear(findScheduleYear(schedule.getSubjectSchedule()));
+        sc.setScheduleStatus(schedule.getScheduleStatus());
+        scheduleRepository.save(sc);
+        return Optional.of(sc);
+    }
+
+    //exception checks for post and put
+    public void exceptionChecks(Schedule schedule){
```
Comment (IamYashwant) [inline src/main/java/com/target/targetreadyresultsservice/service/ScheduleService.java:142]: nice

File: src/main/java/com/target/targetreadyresultsservice/service/SubjectService.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        }
+        return subNames;
+    }
+
+    public  Subject addSubject(Subject subject)  {
+        String code="S_"+subject.getSubjectName().substring(0,2)+subject.getClassCode();
+        subject.setSubjectCode(code);
+        Optional<Subject> savedSub=subjectRepository.findById(subject.getSubjectCode());
+        if(savedSub.isPresent())
+        {   log.info("Subject already present");
```
Comment (IamYashwant) [inline src/main/java/com/target/targetreadyresultsservice/service/SubjectService.java:50]: please take care of the code formatting

File: src/main/java/com/target/targetreadyresultsservice/service/SubjectService.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    public void deleteAll() {
+        subjectRepository.deleteAll();
+    }
+
+    public List<Subject> searchSubjectsByFilters(String subjectName)
+    {
+        log.info("getting subjects with subject name");
+            List<Subject> subjects=subjectRepository.findBySubjectNameIgnoreCase(subjectName);
```
Comment (IamYashwant) [inline src/main/java/com/target/targetreadyresultsservice/service/SubjectService.java:118]: same code formatting

**Review by IamYashwant — CHANGES_REQUESTED** [review]
Few places we have logging and few places its not Ex. check controller 3 doesn't have logging and 2 have. please make logging through. it will help you debug later.

Good to see test cases. nice job there.

**Review by MaheshYadavalli — APPROVED** [review]

**Pull request merged**